pulser-simulation/pulser_simulation/qutip_result.py

```diff
@@ -126,33 +126,64 @@ def _weights(self) -> np.ndarray:
         # Takes care of numerical artefacts in case sum(weights) != 1
         return cast(np.ndarray, weights / sum(weights))
 
+    def _get_relevant_subsystem_indices(
+        self, to_basis: str, tol: float
+    ) -> List[int]:
+        """Finds the subsystems to keep based on the target basis.
+
+        Args:
+            to_basis: The basis to which the state should be transformed.
+
+        Returns:
+            A list of indices representing the relevant subsystems.
+        """
+        if to_basis == "ground-rydberg":
+            ex_state = "2"
+        elif to_basis == "digital":
+            ex_state = "0"
+        else:
+            raise ValueError(
+                "'to_basis' must be 'ground-rydberg' or 'digital'."
+            )
+
+        # Find the subsystems to keep.
+        ex_inds = [
+            i
+            for i in range(3**self._size)
+            if ex_state in np.base_repr(i, base=3).zfill(self._size)
+        ]
+
+        ex_probs = np.abs(self.state.extract_states(ex_inds).full()) ** 2
+        if not np.all(np.isclose(ex_probs, 0, atol=tol)):
+            raise TypeError(
+                "Can't reduce to chosen basis because the population of a "
+                "state to eliminate is above the allowed tolerance."
+            )
+
+        return ex_inds
+
+    def _reduce_to_basis(
+        self, ex_inds: List[int], normalize: bool
+    ) -> qutip.Qobj:
+        """Reduces the state vector by eliminating the states at given indices.
+
+        Args:
+            ex_inds: The indices of the states to eliminate.
+
+        Returns:
+            A Qobj instance representing the state vector with the states eliminated.
+        """
+        state = self.state.copy()
+        state = state.eliminate_states(ex_inds, normalize=normalize)
+        return state
```

Comment:
I'd say this is probably too simple to require being isolated now, I'd move it back to `get_state()`. Also, I think the object is not changed in place so there's no need to copy it